Remove dependency to regex crate

rakus opened the pull request:
Reimplement the `decode_generic` without regular expressions. Just `find` backslashes and process the following characters. 

I used stfu8 in a CLI executable and this resulted in a binary size of ~ 1.8MB. By changing stfu8 to work without regex, the binary size dropped to ~ 700KB.

This also adds two new `DecoderErrorKind` enums: 
* `HexNumberToShort` (not enough characters after a `\x` or `\u`)
* `InvalidHexDigit`

This two errors would previously be reported as `UnescapedSlash`.

This PR also contains a commit for `cargo fmt` and one for `cargo clippy` (SCNR).

rakus committed:
Get rid of regex dependency

Implement decode_generic without regular expression.
Just find backslashes and process following characters.

This removed the dependencies to the regex and lazy_static crates.

vitiral commented on src/decode.rs:
```diff
@@ -47,63 +41,102 @@ pub(crate) enum PushGeneric<'a> {
-            "\\r" => b'\r' as u32,
-            "\\\\" =>b'\\' as u32,
-            "\\x" => from_hex2(&mat.as_str().as_bytes()[2..]) as u32,
+        let (len, c32) = match &string[byte_index..(byte_index + 2)] {
```
just match on the next byte instead of the string.

vitiral commented on src/decode.rs:
```diff
@@ -47,63 +41,102 @@ pub(crate) enum PushGeneric<'a> {
+            "\\\\" => (2, b'\\' as u32),
+            "\\x" => {
+                if rest < 4 {
+                    Err(DecodeError {
```
do a return `Err` and eliminate the `else` clause.

vitiral commented on src/decode.rs:
```diff
@@ -47,63 +41,102 @@ pub(crate) enum PushGeneric<'a> {
+                } else {
+                    (
+                        4,
+                        from_hex2(&string[(byte_index + 2)..].as_bytes()[..2]).map_or_else(
```
again, use a `match` instead of `map_or_else`

vitiral commented on src/decode.rs:
```diff
@@ -47,63 +41,102 @@ pub(crate) enum PushGeneric<'a> {
-                    // to record it's value "as is".
-                    None => c32,
+                if rest < 8 {
+                    Err(DecodeError {
```
return, eliminate `else`

vitiral commented on src/decode.rs:
```diff
@@ -47,63 +41,102 @@ pub(crate) enum PushGeneric<'a> {
+                    })?
+                } else {
+                    let hex6 = &string[(byte_index + 2)..byte_index + 8].as_bytes();
+                    debug_assert_eq!(6, hex6.len());
```
remove?

vitiral commented on src/helpers.rs:
```diff
@@ -6,24 +6,37 @@
 pub(crate) const BSLASH_U16: u16 = BSLASH as u16;
 
 /// create `u8` from two bytes of hex
-pub(crate) fn from_hex2(hex2: &[u8]) -> u8 {
```
can we use `u32::from_str_radix` instead of all these `from_hex` methods?

https://doc.rust-lang.org/std/primitive.u32.html#method.from_str_radix

vitiral commented on src/decode.rs:
```diff
@@ -47,63 +41,102 @@ pub(crate) enum PushGeneric<'a> {
+                            offset += byte_index + 8;
+                            continue;
+                        }
+                        // It is not a valid code point. Still try
```
This code (which was originally mine!) is confusing. Please just do as the others do -- `PushGeneric::Value` followed by `continue`.

I _think_ if you do that you can remove _all_ of the `continue` statements.

vitiral commented on src/encode_u8.rs:
```diff
@@ -25,47 +25,52 @@ pub(crate) fn encode(encoder: &super::Encoder, v: &[u8]) -> String {
+                index += 1;
+                // we needed data, but there was none: error!
+                if index >= len {
+                    index -= 1; // added by me
```
I have no idea what this bit is. Remove all these weird comments.

vitiral reviewed (commented):
I think this is great, thanks for the PR!

I would have appreciated if the formatting was a separate CL first, it would have made reviewing easier. If possible can you split it that way?

rakus commented:
That's why the PR has three commits, but I can split it. Would you like two PRs (1: fmt + clippy, 2: regex) or three (1: fmt, 2: clippy, 3: regex)?
Will also look at your code review comments.

vitiral commented:
> Would you like two PRs (1: fmt + clippy, 2: regex) or three (1: fmt, 2: clippy, 3: regex)?

The former (splitting into just two) is fine, thanks!

rakus commented:
See #8 . Will rebase this PR after merge of #8.

rakus committed:
Reimplement without regex

vitiral commented:
Sorry, I wasn't clear on what I meant regarding removing continues

I meant that the end code that does

```
        push_val(pg)?;
        string = &string[(byte_index + len)..];
        offset += byte_index + len;
```

Should be in every branch. Since it's in every branch, it's not necessary to do a `continue` in any branch (nor do any branch "return" a value).

I don't think we should do `PushGeneric::Char` -- much better that valid characters go onto the currently building string IMO

rakus commented:
I do not understand. The decode function is the generic implementation for `decode_u8` or `decode_u16`.  The `PushGeneric::String` is just used as a transport vehicle that is then either used `as_bytes` to extend the `Vec<u8>` or (character by character) used to extend the `Vec<u16>`.

So for  a valid `\uxxxxxx` with `PushGeneric::String`:
1. decode as char
2. allocate String
3. append Char to String
4. Create `PushGeneric::String`
5. call `push_val` to do a `as_bytes` or `encode_utf16` and append to result vector

With `PushGeneric::Char` this would be:
1. decode as char
2. create `PushGeneric::Char`
3. call `push_val` to do a `encode_utf8` or `encode_utf16` and append to result vector

vitiral commented:
The previous implementation didn't have `PushGeneric::Char`, can we use it's approach of allocating a `String::with_capacity(4)`?

Adding another case will require users (the ones who are writing the `push_val` function) will have to handle the additional case, with the only difference between `String` and `Char` that `Char` is always len=1 whereas string is only sometimes len=1 (not a relevant distinction IMO).

rakus committed:
Implemented with multiple push_val calls

rakus commented:
Ok, now implemented it with multiple push_val calls and removed `PushGeneric::Char`. 
Still advancing to next part of the string at one location (not in every arm of the match), as I think this is more clear and easier to maintain.

Didn't use `String::with_capacity(4)`, but simple `c.to_string()`. 

Just as a new commit, so still needs squashing if you agree to this implementation.

vitiral reviewed (approved):
Looks great, thanks!

I'll merge soon and push a new version.

vitiral commented:
squashed and merged in 6ab362a

vitiral commented:
published to crates.io as version `0.2.7` 06dfc82

vitiral commented:
I forgot to make this clear, but you agree to all the Licensing terms in the LICENSE files (MIT/Apache 2.0), yes?

rakus commented:
Yes I do!
